### PR TITLE
Proxy: Added support for PHP 7 return & parameter type hints

### PR DESCRIPTION
**Pull request by Majkl578**
This PR adds support for PHP 7 return types and parameter type hints on generated proxies.

### CODE REVIEW

File: lib/Doctrine/Common/Proxy/ProxyGenerator.php [hunk 1]
```diff
@@ -777,6 +777,22 @@ private function generateMethods(ClassMetadata $class)
             }
 
             $methods .= $name . '(' . $this->buildParametersString($class, $method, $method->getParameters()) . ')';
+
+            if (PHP_VERSION_ID >= 70000 && $method->hasReturnType()) {
```
Comment (Ocramius) [inline lib/Doctrine/Common/Proxy/ProxyGenerator.php]: Can this be split into a private method?

Comment (Majkl578) [inline lib/Doctrine/Common/Proxy/ProxyGenerator.php]: Done.

**Comment by TomasVotruba** [comment]
:+1:

File: tests/Doctrine/Tests/Common/Proxy/ProxyClassGeneratorTest.php [hunk 2]
```diff
@@ -184,6 +184,52 @@ public function testClassWithVariadicArgumentOnProxiedMethod()
...
+        $this->assertEquals(1, substr_count($classCode, 'function typeHintsWithVariadic(int ...$foo)'));
+    }
+
+    public function testClassWithReturnTypesOnProxiedMethods()
+    {
+        if (PHP_VERSION_ID < 70000) {
+            $this->markTestSkipped('Method return types are only supported in PHP >= 7.0.0.');
+        }
+
+        $className = 'Doctrine\Tests\Common\Proxy\ReturnTypesClass';
```
Comment (TomasVotruba) [inline tests/Doctrine/Tests/Common/Proxy/ProxyClassGeneratorTest.php]: This can use PHP 5.6 `::class`, right?

Comment (Majkl578) [inline tests/Doctrine/Tests/Common/Proxy/ProxyClassGeneratorTest.php]: No, that is a syntax error on PHP <5.5 (and this package is >=5.3.2).

Comment (TomasVotruba) [inline tests/Doctrine/Tests/Common/Proxy/ProxyClassGeneratorTest.php]: This code is never run on PHP 5.6 and lower. Or is there some linter?

Comment (Ocramius) [inline tests/Doctrine/Tests/Common/Proxy/ProxyClassGeneratorTest.php]: It would fail at parse time, fyi ;-)

Comment (TomasVotruba) [inline tests/Doctrine/Tests/Common/Proxy/ProxyClassGeneratorTest.php]: I realized, to bad :/

**Comment by Majkl578** [comment]
@Ocramius: Is there anything else that needs to be done?

**Comment by Ocramius** [comment]
@Majkl578 no, this looks good, but the patch has to wait for 2.6.0, so I'll first have to branch out 2.5.x to keep it stable.

**Comment by Majkl578** [comment]
@Ocramius: Makes sense. :+1: 

*(Hoping that 2.6 will be released when PHP 7 reaches stable release. :) )*

File: tests/Doctrine/Tests/Common/Proxy/ProxyClassGeneratorTest.php (hunk 2 above)
Comment (Ocramius) [inline tests/Doctrine/Tests/Common/Proxy/ProxyClassGeneratorTest.php]: Gonna change it anyway: bumping the library dependencies.

Comment (Majkl578) [inline tests/Doctrine/Tests/Common/Proxy/ProxyClassGeneratorTest.php]: Cool, let me know if/when you need rebase or other changes here.

**Comment by Ocramius** [comment]
Can haz PHP 7 support plz? Yes, we can, thanks to @Majkl578 :-)

**Pull request closed without merging**